src/raw/mod.rs

```diff
@@ -192,14 +192,40 @@ impl ProbeSeq {
 // Workaround for emscripten bug emscripten-core/emscripten-fastcomp#258
 #[cfg_attr(target_os = "emscripten", inline(never))]
 #[cfg_attr(not(target_os = "emscripten"), inline)]
-fn capacity_to_buckets(cap: usize) -> Option<usize> {
+fn capacity_to_buckets(cap: usize, table_layout: TableLayout) -> Option<usize> {
     debug_assert_ne!(cap, 0);
 
+    // todo: what happens with ZSTs?
+    // Consider a small layout like TableLayout { size: 1, ctrl_align: 16 } on
+    // a platform with Group::WIDTH of 16 (like x86_64 with SSE2). For small
+    // bucket sizes, this ends up wasting quite a few bytes just to pad to the
+    // relatively larger ctrl_align:
+    //
+    // | capacity | buckets | bytes allocated | bytes per item |
+    // | -------- | ------- | --------------- | -------------- |
+    // |        3 |       4 |              36 | (Yikes!)  12.0 |
+    // |        7 |       8 |              40 | (Poor)     5.7 |
+    // |       14 |      16 |              48 |            3.4 |
+    // |       28 |      32 |              80 |            3.3 |
+    //
+    // The ratio of ctrl_align / size is used to set a minimum buckets so
+    // that padding to the alignment isn't dominating the total allocation.
+    let cap = {
+        let q = table_layout.ctrl_align / table_layout.size.max(1);
+        match q.checked_mul(7) {
+            None => cap,
+            Some(x) => {
+                let adjusted = x / 8;
+                adjusted.max(cap)
+            }
+        }
+    };
```

Comment:
Rather than all this complicated logic, this could all be moved as hard-coded logic into the `cap < 8` case below. Specifically, when `cap <= 7` then we need to choose a bucket count of 4, 8 or 16 depending on the values of `size_of::<T>()` and `cap`.

Reply:
`cap < 8` isn't sufficient for `size_of::<T>() == 1` on platforms with a group size of 16 as `buckets * table_layout.size >= table.ctrl_align` isn't true e.g. `8 * 1 >= 16`. That's why there's more complication.

With that said, with fresh eyes I don't like how the PR is currently written 😄. I'll take a look at improving it next week.

Reply:
For now I at least adjusted the code comments and names to be a bit more explanatory.

Reply:
I found something that seems better to me, at least.